server/src/main/java/org/elasticsearch/action/admin/cluster/desiredbalance/TransportGetDesiredBalanceAction.java

```diff
@@ -0,0 +1,132 @@
+import org.elasticsearch.cluster.routing.allocation.allocator.DesiredBalanceShardsAllocator;
+import org.elasticsearch.cluster.routing.allocation.allocator.ShardAssignment;
+import org.elasticsearch.cluster.service.ClusterService;
+import org.elasticsearch.common.inject.Inject;
+import org.elasticsearch.tasks.Task;
+import org.elasticsearch.threadpool.ThreadPool;
+import org.elasticsearch.transport.TransportService;
+
+import java.util.HashMap;
+import java.util.Map;
+
+public class TransportGetDesiredBalanceAction extends TransportMasterNodeReadAction<DesiredBalanceRequest, DesiredBalanceResponse> {
+
+    private final DesiredBalanceShardsAllocator desiredBalanceShardsAllocator;
+
+    @Inject
+    public TransportGetDesiredBalanceAction(
+        TransportService transportService,
+        ClusterService clusterService,
+        ThreadPool threadPool,
+        ActionFilters actionFilters,
+        IndexNameExpressionResolver indexNameExpressionResolver,
+        DesiredBalanceShardsAllocator desiredBalanceShardsAllocator
+    ) {
+        super(
+            GetDesiredBalanceAction.NAME,
+            transportService,
+            clusterService,
+            threadPool,
+            actionFilters,
+            DesiredBalanceRequest::new,
+            indexNameExpressionResolver,
+            DesiredBalanceResponse::from,
+            ThreadPool.Names.SAME
+        );
+        this.desiredBalanceShardsAllocator = desiredBalanceShardsAllocator;
+    }
+
+    @Override
+    protected void masterOperation(
+        Task task,
+        DesiredBalanceRequest request,
+        ClusterState state,
+        ActionListener<DesiredBalanceResponse> listener
+    ) throws Exception {
+        String allocatorName = ClusterModule.SHARDS_ALLOCATOR_TYPE_SETTING.get(state.metadata().settings());
+        if (allocatorName.equals(ClusterModule.DESIRED_BALANCE_ALLOCATOR) == false) {
+            listener.onFailure(
+                new ResourceNotFoundException(
+                    "Expected the shard balance allocator to be `desired_balance`, but got `" + allocatorName + "`"
+                )
+            );
+            return;
+        }
+
+        DesiredBalance latestDesiredBalance = desiredBalanceShardsAllocator.getDesiredBalance();
+        if (latestDesiredBalance == null) {
+            listener.onFailure(new ResourceNotFoundException("Desired balance is not computed yet"));
+            return;
+        }
+        Map<String, Map<Integer, DesiredBalanceResponse.DesiredShards>> routingTable = new HashMap<>();
+        for (IndexRoutingTable indexRoutingTable : state.routingTable()) {
+            Map<Integer, DesiredBalanceResponse.DesiredShards> indexDesiredShards = new HashMap<>();
+            for (int shardId = 0; shardId < indexRoutingTable.size(); shardId++) {
+                IndexShardRoutingTable shardRoutingTable = indexRoutingTable.shard(shardId);
+                // Do replicas overwrite primary shards here? Maybe we should we return
```

Comment:
I kind of think that we do need to change the JSON output to return the primary along with replicas for every shard id.